Introduce BackendSelect dispatch key

izdeby opened the pull request:
This stack is the first step toward an effort to fix, clean up and simplify code generation logic.  Please see the master [task](https://github.com/pytorch/pytorch/issues/30405) see related discussions and all the known issues.

The main focus of these changes is TensorOptions in code generation. 
Goals:
- Remove TensorOptions from generated code wherever it's possible. Leave it only in python/C++ API layers.
- Refactor TensorOptions logic to a single place.
- Log all discovered issues.

Non-goals:
- Fix Everything! 
- Remove all the hacks in code generation scripts.
- Clean up and defector all code generation scripts.

-----------
In this PR:
Introduce BackendSelect dispatch key with a high priority. All factory methods will be registering its special backend selector method with the ops schema so that the dispatcher would navigate there after Variable. Dispatch key id will be calculated and the correct kernel will be called. 

[Generated file](https://gist.github.com/izdeby/8056bf72b50b7c66075d37f334d84bf5)

-----------
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #30983 [WIP] Updated codegen to remove TensorOptions from generated code
* #30982 Prepare templates
* **#30981 Introduce BackendSelect dispatch key**

Differential Revision: [D18912684](https://our.internmc.facebook.com/intern/diff/D18912684)

izdeby committed:
Updated DispatchKeyExtractor to expect TensorOptions

[ghstack-poisoned]

gchanan commented on aten/src/ATen/core/dispatch/DispatchKeyExtractor.h:
```diff
@@ -48,11 +48,65 @@ namespace detail {
+  };
+
+  template<class Arg> using arg_is_tensor_option_arg = guts::typelist::contains<
+    guts::typelist::typelist<c10::optional<ScalarType>, c10::optional<Layout>, 
```
Also, doesn't this allow more than you want?  I.e. I can have mixtures of optional and non-optional stuff, even optional and non-optional versions of the same property.

It seems saner to just split these up into two lists.

gchanan reviewed (commented):
I don't really understand why we need this.  Why should the dispatcher even know about TensorOptions?  Can't it just dispatch to the non-TensorOptions variants of the factory functions?

izdeby commented:
> I don't really understand why we need this. Why should the dispatcher even know about TensorOptions? Can't it just dispatch to the non-TensorOptions variants of the factory functions?

In order to figure out which variant to pick we need to calculate the type_set. Thats done in [TensorOptions.h](https://github.com/pytorch/pytorch/blob/master/c10/core/TensorOptions.h#L339-L343) and requires layout and device. We can refactor it later.

izdeby committed:
Update on "Updated DispatchKeyExtractor to expect TensorOptions"

This stack is a first step toward an effort to fix, clean up and simplify code generation logic.  Please see the master [task](https://github.com/pytorch/pytorch/issues/30405) to see related discussions and all the known issues.

Main focus of these changes is TensorOptions in code generation. 
Goals:
- Remove TensorOptions from generated code wherever it's possible. Leave it only in python/C++ API layers.
- Refactor TensorOptions logic to a single place.
- Log all discovered issues.

Non goals:
- Fix Everything! 
- Remove all the hacks in code generation scripts.
- Clean up and defector all code generation scripts.

-----------
In this PR:
Extended DispatchKeyExtractor logic to expect TensorOptions.

-----------



[ghstack-poisoned]

smessmer commented on aten/src/ATen/core/dispatch/DispatchKeyExtractor.h:
```diff
@@ -48,13 +48,87 @@ namespace detail {
+
+  // NOTE: [schema mismatch in native_functions.yaml]
+  // Due to this issue we have to check for two sets of TensorOptions types:
+  // optinal and non-optional.
```
nit: typo

smessmer commented on aten/src/ATen/core/dispatch/DispatchKeyExtractor.h:
```diff
@@ -48,13 +48,87 @@ namespace detail {
+                              c10::optional<bool>>,
+     guts::remove_const_t<guts::remove_reference_t<Arg>>>;
+
+  template<class... Args> using args_have_tensor_options = guts::disjunction<
```
We need to have the checks we discussed here, i.e.

return `args_have_tensor_options == true` if and only if all of the following are true
- it has either ScalarType, Layout, Device and bool arguments
- either all of them are non-optional or all of them are optional
- they are specified in exactly that order without other arguments inbetween

What this code currently is doing is it returns  `args_have_tensor_options == true` as soon as any argument is `ScalarType`, `Layout`, `Device`, `bool` or optional of any of these types. So, for example, this code will interpret any operator with any bool argument as `TensorOptions`. This is wrong.

smessmer commented on aten/src/ATen/core/dispatch/DispatchKeyExtractor.h:
```diff
@@ -48,13 +48,87 @@ namespace detail {
+    void operator()(Layout layout) {
+      options = options.layout(layout);
+    }
+    void operator()(bool pin_memory) {
```
this is dangerous. Even if the operator has the correct set and order of TensorOptions args, there might be additional bool arguments and they will go into here even though they're not `pin_memory`. You need to make sure that only the `bool` argument that comes directly after the `Layout` argument goes into here. This probably means that you cannot use `IterArgs` anymore but need to write your own metaprogramming.

izdeby committed:
Update on "Updated DispatchKeyExtractor to expect TensorOptions"

This stack is a first step toward an effort to fix, clean up and simplify code generation logic.  Please see the master [task](https://github.com/pytorch/pytorch/issues/30405) to see related discussions and all the known issues.

Main focus of these changes is TensorOptions in code generation. 
Goals:
- Remove TensorOptions from generated code wherever it's possible. Leave it only in python/C++ API layers.
- Refactor TensorOptions logic to a single place.
- Log all discovered issues.

Non goals:
- Fix Everything! 
- Remove all the hacks in code generation scripts.
- Clean up and defector all code generation scripts.

-----------
In this PR:
Extended DispatchKeyExtractor logic to expect TensorOptions.

-----------

Differential Revision: [D18912684](https://our.internmc.facebook.com/intern/diff/D18912684)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/core/dispatch/DispatchKeyExtractor.h:
```diff
@@ -48,13 +48,87 @@ namespace detail {
+    }
+    void operator()(c10::optional<Device> device) {
+      if (device.has_value()) {
+        options = options.device(*device);
```
I'm pretty sure the setters all take `optional`, so you can just bang `device` in directly here without doing a test here...

ezyang commented on aten/src/ATen/core/dispatch/DispatchKeyExtractor.h:
```diff
@@ -48,11 +48,65 @@ namespace detail {
+      if (dtype.has_value())
+        options = options.dtype(*dtype);
+      else 
+        options = options.dtype(at::get_default_dtype());
```
I'm a little confused here: if you're populating TensorOptions here, why do you need to fill it in explicitly with default dtype? If you use the normal getter to extract values from tensor options, you'll get the right behavior in any case, no?

izdeby committed:
Update on "Updated DispatchKeyExtractor to expect TensorOptions"

This stack is a first step toward an effort to fix, clean up and simplify code generation logic.  Please see the master [task](https://github.com/pytorch/pytorch/issues/30405) to see related discussions and all the known issues.

Main focus of these changes is TensorOptions in code generation. 
Goals:
- Remove TensorOptions from generated code wherever it's possible. Leave it only in python/C++ API layers.
- Refactor TensorOptions logic to a single place.
- Log all discovered issues.

Non goals:
- Fix Everything! 
- Remove all the hacks in code generation scripts.
- Clean up and defector all code generation scripts.

-----------
In this PR:
Extended DispatchKeyExtractor logic to expect TensorOptions.

-----------

Differential Revision: [D18912684](https://our.internmc.facebook.com/intern/diff/D18912684)

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit 45c1740:
* **1/1** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=45c17409e4227dcf0016d2ef63929e478667fa40).

### 1 new failure recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3935070) pytorch-linux-xenial-py3-clang5-android-ndk-r19c-gradle-build (1/1)
**Step:** "pytorch android gradle build" ([details](https://dr.pytorch.org/build-details.html?build_id=73658186))
```
sys	0m0.158s 
++ docker run --cap-add=SYS_PTRACE --security-opt seccomp=unconfined -t -d -w /var/lib/jenkins 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3-clang5-android-ndk-r19c:405-45c17409e4227dcf0016d2ef63929e478667fa40-android-x86_32 
+ export id_x86_32=80ea794046fe0889762de1a78981d95336218962ce964c90dfe9337f9daa3079 
+ id_x86_32=80ea794046fe0889762de1a78981d95336218962ce964c90dfe9337f9daa3079 
+ export 'COMMAND=((echo "source ./workspace/env" && echo "sudo chown -R jenkins workspace") | docker exec -u jenkins -i "$id_x86_32" bash) 2>&1' 
+ COMMAND='((echo "source ./workspace/env" && echo "sudo chown -R jenkins workspace") | docker exec -u jenkins -i "$id_x86_32" bash) 2>&1' 
+ echo '((echo' '"source' './workspace/env"' '&&' echo '"sudo' chown -R jenkins 'workspace")' '|' docker exec -u jenkins -i '"$id_x86_32"' 'bash)' '2>&1' 
+ unbuffer bash ./command.sh 
+ ts 
+ docker pull 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3-clang5-android-ndk-r19c:405-45c17409e4227dcf0016d2ef63929e478667fa40-android-arm-v7a 
Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3-clang5-android-ndk-r19c:405-45c17409e4227dcf0016d2ef63929e478667fa40-android-arm-v7a not found 
```


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

izdeby committed:
Update on "Updated DispatchKeyExtractor to expect TensorOptions"

This stack is a first step toward an effort to fix, clean up and simplify code generation logic.  Please see the master [task](https://github.com/pytorch/pytorch/issues/30405) to see related discussions and all the known issues.

Main focus of these changes is TensorOptions in code generation. 
Goals:
- Remove TensorOptions from generated code wherever it's possible. Leave it only in python/C++ API layers.
- Refactor TensorOptions logic to a single place.
- Log all discovered issues.

Non goals:
- Fix Everything! 
- Remove all the hacks in code generation scripts.
- Clean up and defector all code generation scripts.

-----------
In this PR:
Extended DispatchKeyExtractor logic to expect TensorOptions.

-----------

Differential Revision: [D18912684](https://our.internmc.facebook.com/intern/diff/D18912684)

[ghstack-poisoned]

izdeby committed:
Update on "Updated DispatchKeyExtractor to expect TensorOptions"

This stack is a first step toward an effort to fix, clean up and simplify code generation logic.  Please see the master [task](https://github.com/pytorch/pytorch/issues/30405) to see related discussions and all the known issues.

Main focus of these changes is TensorOptions in code generation. 
Goals:
- Remove TensorOptions from generated code wherever it's possible. Leave it only in python/C++ API layers.
- Refactor TensorOptions logic to a single place.
- Log all discovered issues.

Non goals:
- Fix Everything! 
- Remove all the hacks in code generation scripts.
- Clean up and defector all code generation scripts.

-----------
In this PR:
Extended DispatchKeyExtractor logic to expect TensorOptions.

-----------

Differential Revision: [D18912684](https://our.internmc.facebook.com/intern/diff/D18912684)

izdeby committed:
Update on "Updated DispatchKeyExtractor to expect TensorOptions"

This stack is a first step toward an effort to fix, clean up and simplify code generation logic.  Please see the master [task](https://github.com/pytorch/pytorch/issues/30405) to see related discussions and all the known issues.

Main focus of these changes is TensorOptions in code generation. 
Goals:
- Remove TensorOptions from generated code wherever it's possible. Leave it only in python/C++ API layers.
- Refactor TensorOptions logic to a single place.
- Log all discovered issues.

Non goals:
- Fix Everything! 
- Remove all the hacks in code generation scripts.
- Clean up and defector all code generation scripts.

-----------
In this PR:
Extended DispatchKeyExtractor logic to expect TensorOptions.

-----------

Differential Revision: [D18912684](https://our.internmc.facebook.com/intern/diff/D18912684)

[ghstack-poisoned]

izdeby committed:
Update on "Updated DispatchKeyExtractor to expect TensorOptions"

This stack is a first step toward an effort to fix, clean up and simplify code generation logic.  Please see the master [task](https://github.com/pytorch/pytorch/issues/30405) to see related discussions and all the known issues.

Main focus of these changes is TensorOptions in code generation. 
Goals:
- Remove TensorOptions from generated code wherever it's possible. Leave it only in python/C++ API layers.
- Refactor TensorOptions logic to a single place.
- Log all discovered issues.

Non goals:
- Fix Everything! 
- Remove all the hacks in code generation scripts.
- Clean up and defector all code generation scripts.

-----------
In this PR:
Extended DispatchKeyExtractor logic to expect TensorOptions.

-----------

Differential Revision: [D18912684](https://our.internmc.facebook.com/intern/diff/D18912684)

[ghstack-poisoned]

izdeby committed:
Update on "[WIP] Updated DispatchKeyExtractor to expect TensorOptions"

This stack is a first step toward an effort to fix, clean up and simplify code generation logic.  Please see the master [task](https://github.com/pytorch/pytorch/issues/30405) to see related discussions and all the known issues.

Main focus of these changes is TensorOptions in code generation. 
Goals:
- Remove TensorOptions from generated code wherever it's possible. Leave it only in python/C++ API layers.
- Refactor TensorOptions logic to a single place.
- Log all discovered issues.

Non goals:
- Fix Everything! 
- Remove all the hacks in code generation scripts.
- Clean up and defector all code generation scripts.

-----------
In this PR:
Extended DispatchKeyExtractor logic to expect TensorOptions.

-----------

Differential Revision: [D18912684](https://our.internmc.facebook.com/intern/diff/D18912684)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/BackendSelectRegister.cpp:
```diff
@@ -0,0 +1,631 @@
+}
+
+Tensor empty_like(const Tensor & self, const TensorOptions & options, c10::optional<MemoryFormat> memory_format) {
+  return at::native::empty_like(self, options, memory_format);
```
Hmm?

izdeby replied:
removed

ezyang commented on aten/src/ATen/native/BackendSelectRegister.cpp:
```diff
@@ -0,0 +1,631 @@
+#include <ATen/Dispatch.h>
+#include <ATen/core/op_registration/op_registration.h>
+#include <c10/core/TensorOptions.h>
+
```
It would be good to have a comment here explaining the function of this file. I also wonder a bit about whether or not we should do things to reduce duplication here. For example, the types for each of these functions has to be written twice (the function definition, and then the template argument) and it would be great if we only had to write it once. 

@smessmer, if I have a template function `template <T> f(T&&)` and I call it from some other function `g(S x) { 
f(x); }`, is the `T` in `f` guaranteed to deduce to `S` no matter what S is?

smessmer replied:
Basic rule: `&` and `&&` together decay to `&`.

Given a non-reference concrete type `A`.
- If `S` is a lvalue reference (i.e. `S=A&`), then `T=S=A&`.
- If `S` is a non-reference (i.e. `S=A`), then `T=S&=A&` because a non-reference `A` can be passed to functions taking `A` or `A&` but not to functions taking `A&&`. `T=S` would lead to `T&& = A&&` and the argument type wouldn't match the call site. So instead, `T=S&=A&` and `T&& = (A&)&& = A&` decays, the argument can be passed in.
- If `S` is an rvalue reference (i.e. `S=A&&`), then `T=decay_t<S>=A`. This allows `T&& = A&&` and allows moving in the argument.

Note that this assumes that your call to `f(x)` forwards, i.e. looks like `f(std::forward<S>(x))`. In the way you've written it, you're always passing `x` by value, which means `T` will always be `S&`.

smessmer replied:
For simplifying this code here, the easiest way is probably to use a template and specify the type arguments explicitly instead of inferring them.

ezyang replied:
Hmm, but I don't think we ever have `A&&` arguments today, so maybe it would work out?

gchanan replied:
why do callUnboxedWithDispatchKey templates not match the decltype of the function?  Is it because we want to apply some additional checking?

izdeby replied:
> why do callUnboxedWithDispatchKey templates not match the decltype of the function? Is it because we want to apply some additional checking?

It does match. It just looks a bit different due to how callUnboxedWithDispatchKey is defined:

  ```
template<class Return, class... Args>

 Return callUnboxedWithDispatchKey(const OperatorHandle& op, DispatchKey dispatchKey, Args... args) const;

```

So when we have a call like an `eye`
`op.callUnboxedWithDispatchKey<Tensor, int64_t, const TensorOptions &>(key, n, options);`
`Tensor` - `Return`
`int64_t, const TensorOptions & `- `Args`
`key` - `DispatchKey dispatchKey`
`n, options`- `Args... args`

izdeby replied:
> It would be good to have a comment here explaining the function of this file. I also wonder a bit about whether or not we should do things to reduce duplication here. For example, the types for each of these functions has to be written twice (the function definition, and then the template argument) and it would be great if we only had to write it once.
> 
> @smessmer, if I have a template function `template <T> f(T&&)` and I call it from some other function `g(S x) { f(x); }`, is the `T` in `f` guaranteed to deduce to `S` no matter what S is?

Added a comment

smessmer commented on c10/core/DispatchKey.cpp:
```diff
@@ -40,6 +40,8 @@ const char* toString(DispatchKey t) {
       return "ComplexCUDATensorId";
     case DispatchKey::VariableTensorId:
       return "VariableTensorId";
+    case DispatchKey::BackendSelect:
```
Why are we calling this `BackendSelect` and not `CompoundOp`? This should be the same mechanism, right?

smessmer replied:
nvm, saw your comment in the doc

smessmer commented on c10/core/DispatchKeySet.h:
```diff
@@ -122,7 +122,7 @@ C10_API std::ostream& operator<<(std::ostream&, DispatchKeySet);
 // have to adjust this some more (sorry.)
 static inline DispatchKey legacyExtractDispatchKey(DispatchKeySet s) {
-  return s.remove(DispatchKey::VariableTensorId).highestPriorityTypeId();
+  return s.remove(DispatchKey::VariableTensorId).remove(DispatchKey::BackendSelect).highestPriorityTypeId();
```
is this a perf hit? Should we combine the removal of both into one call?

ezyang replied:
legacy extractor isn't perf critical (or, if it is, we should get rid of it and fix it up)

smessmer reviewed (commented):
Shouldn't we add a fallthrough for this key somewhere?

izdeby committed:
Update on "Introduce BackendSelect dispatch key"

This stack is the first step toward an effort to fix, clean up and simplify code generation logic.  Please see the master [task](https://github.com/pytorch/pytorch/issues/30405) see related discussions and all the known issues.

The main focus of these changes is TensorOptions in code generation. 
Goals:
- Remove TensorOptions from generated code wherever it's possible. Leave it only in python/C++ API layers.
- Refactor TensorOptions logic to a single place.
- Log all discovered issues.

Non-goals:
- Fix Everything! 
- Remove all the hacks in code generation scripts.
- Clean up and defector all code generation scripts.

-----------
In this PR:
Introduce BackendSelect dispatch key with a high priority. All factory methods will be registering its special backend selector method with the ops schema so that the dispatcher would navigate there after Variable. Dispatch key id will be calculated and the correct kernel will be called. 



-----------

Differential Revision: [D18912684](https://our.internmc.facebook.com/intern/diff/D18912684)

[ghstack-poisoned]

gchanan commented on aten/src/ATen/native/BackendSelectRegister.cpp:
```diff
@@ -0,0 +1,615 @@
+  return op.callUnboxedWithDispatchKey<Tensor, IntArrayRef, IntArrayRef, const TensorOptions &>(key, size, stride, options);
+}
+
+Tensor empty_affine_quantized(IntArrayRef size, const TensorOptions & options, double scale, int64_t zero_point, c10::optional<MemoryFormat> memory_format) {
```
shouldn't this be _empty_affine_quantized?

izdeby replied:
renamed

bhosmer commented on aten/src/ATen/native/BackendSelectRegister.cpp:
```diff
@@ -0,0 +1,615 @@
+  .op(torch::RegisterOperators::options()
+    .schema("aten::triu_indices(int row, int col, int offset=0, *, ScalarType? dtype=long, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor")
+    .impl_unboxedOnlyKernel<decltype(triu_indices), &triu_indices>(DispatchKey::BackendSelect)
+    .aliasAnalysis(c10::AliasAnalysisKind::FROM_SCHEMA));
```
IIUC this duplicates a big block of schema declarations from `native_functions.yaml`, right? If so, that seems like a pretty nontrivial downgrade to the process of adding/modifying an op, no? You have to be aware that this new slab of duplicate schema-bearing declarations exists, understand what it does

First just because it's a big redundant slab of schema-bearing declarations that have to be known to exist and manually kept in sync, second because its relationship to native_functions isn't clear - does your new op belong here or no?)

Unless I'm missing something this feels like an instance of the kind of thing we're trying to remove from the process, right? If it's really necessary, it seems like the only way to avoid the new wrinkle is to bring this code under codegen, right?

bhosmer replied:
Codegen or not, it's going to be much easier to read and/or write these if implementation and registration are colocated, e.g. 
```
Tensor full(IntArrayRef size, Scalar fill_value, const TensorOptions & options) {
  DispatchKey key = options.computeDispatchKey();
  static auto op = c10::Dispatcher::singleton().findSchemaOrThrow("aten::full", "");
  return op.callUnboxedWithDispatchKey<Tensor, IntArrayRef, Scalar, const TensorOptions &>(key, size, fill_value, options);
}

auto full_options = torch::RegisterOperators::options()
    .schema("aten::full(int[] size, Scalar fill_value, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor")
    .impl_unboxedOnlyKernel<decltype(full), &full>(DispatchKey::BackendSelect)
    .aliasAnalysis(c10::AliasAnalysisKind::FROM_SCHEMA);

// then later

static auto registry = torch::RegisterOperators()
  ...
  .op(full_options)
  ...

```
...or by putting an implementation lambda directly in the registry, or whatever

bhosmer replied:
...but again the larger issue is that adding these decl blocks as a second world, redundant to but untethered from native_functions.yaml, just seems like a recipe for more devex friction. Is that POV off base? 

Going back now to better understand Greg's "why do we need this at all?" comment...

izdeby replied:
I agree on the codegen part, but @gchanan did seem too excited about this idea. it would make stuff easier to maintain and support. 

But I disagree with the redundancy of this. We have discussed this during the c10 and the follow-up meetings. we need this for dispatching, especially once TensorOptions will be gone, this code will become crucial for dispatching.

I can rewrite it following your example, no problems. We just have to agree if we want to codegen this or not.

ezyang replied:
Part of the trouble is there is an ordering issue, where the API for registration is not as nice as it ought to be. For example, if it were possible to write registrations without having to put in the entire schema string, that would reduce the duplication quite a bit. But the work in that area isn't here yet.

Because there is an *intention* to change this API, and we may change the API multiple times, my personal thinking is that we should codegen the registrations, because it is less typing to modify a single template, than to have to modify 20 or so hand written occurrences.

izdeby replied:
Ok, will update the PR to codegen this.

bhosmer replied:
SGTM but were you saying @gchanan wasn't digging that idea? Interested what his POV is here

ezyang replied:
gchanan, izdeby and I had a call and gchanan is OK with the codegen, as it is simpler codegen than what we had before.

izdeby committed:
Update on "Introduce BackendSelect dispatch key"

This stack is the first step toward an effort to fix, clean up and simplify code generation logic.  Please see the master [task](https://github.com/pytorch/pytorch/issues/30405) see related discussions and all the known issues.

The main focus of these changes is TensorOptions in code generation. 
Goals:
- Remove TensorOptions from generated code wherever it's possible. Leave it only in python/C++ API layers.
- Refactor TensorOptions logic to a single place.
- Log all discovered issues.

Non-goals:
- Fix Everything! 
- Remove all the hacks in code generation scripts.
- Clean up and defector all code generation scripts.

-----------
In this PR:
Introduce BackendSelect dispatch key with a high priority. All factory methods will be registering its special backend selector method with the ops schema so that the dispatcher would navigate there after Variable. Dispatch key id will be calculated and the correct kernel will be called. 



-----------

Differential Revision: [D18912684](https://our.internmc.facebook.com/intern/diff/D18912684)

[ghstack-poisoned]

izdeby committed:
Update on "Introduce BackendSelect dispatch key"

This stack is the first step toward an effort to fix, clean up and simplify code generation logic.  Please see the master [task](https://github.com/pytorch/pytorch/issues/30405) see related discussions and all the known issues.

The main focus of these changes is TensorOptions in code generation. 
Goals:
- Remove TensorOptions from generated code wherever it's possible. Leave it only in python/C++ API layers.
- Refactor TensorOptions logic to a single place.
- Log all discovered issues.

Non-goals:
- Fix Everything! 
- Remove all the hacks in code generation scripts.
- Clean up and defector all code generation scripts.

-----------
In this PR:
Introduce BackendSelect dispatch key with a high priority. All factory methods will be registering its special backend selector method with the ops schema so that the dispatcher would navigate there after Variable. Dispatch key id will be calculated and the correct kernel will be called. 



-----------

Differential Revision: [D18912684](https://our.internmc.facebook.com/intern/diff/D18912684)

[ghstack-poisoned]

izdeby committed:
Update on "Introduce BackendSelect dispatch key"

This stack is the first step toward an effort to fix, clean up and simplify code generation logic.  Please see the master [task](https://github.com/pytorch/pytorch/issues/30405) see related discussions and all the known issues.

The main focus of these changes is TensorOptions in code generation. 
Goals:
- Remove TensorOptions from generated code wherever it's possible. Leave it only in python/C++ API layers.
- Refactor TensorOptions logic to a single place.
- Log all discovered issues.

Non-goals:
- Fix Everything! 
- Remove all the hacks in code generation scripts.
- Clean up and defector all code generation scripts.

-----------
In this PR:
Introduce BackendSelect dispatch key with a high priority. All factory methods will be registering its special backend selector method with the ops schema so that the dispatcher would navigate there after Variable. Dispatch key id will be calculated and the correct kernel will be called. 



-----------

Differential Revision: [D18912684](https://our.internmc.facebook.com/intern/diff/D18912684)

[ghstack-poisoned]

bhosmer commented on aten/src/ATen/gen_backend_select_register.py:
```diff
@@ -0,0 +1,47 @@
+}
+""")
+
+def register_backend_select_methods(declarations, top_env):
```
I think it'd be more straightforward if this function simply took declarations and returned generated artifacts, rather than also taking a dictionary and adding the artifacts to it.

I know this pattern is easy to fall into because of the convention of calling `CodeTemplate.substitute()` with an environment, but heavy use of dynamic dictionaries instead of explicit params is one of the patterns that makes the codegen pipeline so hard to reason about - take a look at `gen_python_functions` now, it no longer uses env dictionaries at all, even in the calls to CodeTemplate.substitute(). I think it makes it quite a bit easier to track what's going (at least without intensive study :)

izdeby replied:
Yeah, i totally agree. Good call. 
Updated

bhosmer commented on aten/src/ATen/templates/BackendSelectRegister.cpp:
```diff
@@ -0,0 +1,22 @@
+${backend_select_method_definitions}
+
+static auto registry = torch::RegisterOperators()
+  ${backend_select_function_registrations};
```
We were going to colocate the definitions with registrations, right?

izdeby replied:
yes

bhosmer reviewed (changes requested):
One little thing and one big but should be easy thing and it'll be good to go from my POV :)

izdeby committed:
Update on "[WIP] Introduce BackendSelect dispatch key"

This stack is the first step toward an effort to fix, clean up and simplify code generation logic.  Please see the master [task](https://github.com/pytorch/pytorch/issues/30405) see related discussions and all the known issues.

The main focus of these changes is TensorOptions in code generation. 
Goals:
- Remove TensorOptions from generated code wherever it's possible. Leave it only in python/C++ API layers.
- Refactor TensorOptions logic to a single place.
- Log all discovered issues.

Non-goals:
- Fix Everything! 
- Remove all the hacks in code generation scripts.
- Clean up and defector all code generation scripts.

-----------
In this PR:
Introduce BackendSelect dispatch key with a high priority. All factory methods will be registering its special backend selector method with the ops schema so that the dispatcher would navigate there after Variable. Dispatch key id will be calculated and the correct kernel will be called. 



-----------

Differential Revision: [D18912684](https://our.internmc.facebook.com/intern/diff/D18912684)

[ghstack-poisoned]

bhosmer commented:
What's here LGTM but I think I might not have been clear re colocating defs and registrations - I mean colocate them pairwise, so e.g. here's how `empty` and `empty_names` (to choose two random ops that are next to each other in the original, non-CG commit) would appear:
```
// in the middle of the list of op/options pairs:

Tensor empty(IntArrayRef size, const TensorOptions & options, c10::optional<MemoryFormat> memory_format) {
  static auto op = c10::Dispatcher::singleton().findSchemaOrThrow("aten::empty", "memory_format");
  DispatchKey key = options.computeDispatchKey();
  return op.callUnboxedWithDispatchKey
<Tensor, IntArrayRef, const TensorOptions &, c10::optional<MemoryFormat>>
(key, size, options, memory_format);
}

static auto empty_options = torch::RegisterOperators::options()
    .schema("aten::empty.memory_format(int[] size, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None, MemoryFormat? memory_format=None) -> Tensor")
    .impl_unboxedOnlyKernel<decltype(empty), &empty>(DispatchKey::BackendSelect)
    .aliasAnalysis(AliasAnalysisKind::FROM_SCHEMA);

Tensor empty_names(IntArrayRef size, c10::optional<DimnameList> names, const TensorOptions & options, c10::optional<MemoryFormat> memory_format) {
  return at::native::empty(size, names, options, memory_format);
}

static auto empty_names_options = torch::RegisterOperators::options()
    .schema("aten::empty.names(int[] size, *, Dimname[]? names, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None, MemoryFormat? memory_format=None) -> Tensor")
    .impl_unboxedOnlyKernel<decltype(empty_names), &empty_names>(DispatchKey::BackendSelect)
    .aliasAnalysis(c10::AliasAnalysisKind::FROM_SCHEMA);

// then a bunch more op/options pairs, then: 

static auto registry = torch::RegisterOperators()
  ...
  .op(empty_options)
  .op(empty_names_options)
  ...
```

...there are other ways to group the pairs together, this was just a way that seemed readable to me. But the main thing is that all the nontrivial stuff related to op Foo appear in one place. (Even putting this example together drove the point home of how much friction it adds to have to go hunting around 😄) 

Also FWIW, in this example I swapped the first two lines in the function def and added line breaks to the `callUnboxedWithDispatchKey` call - just some ideas for cheap ways to add readability.

LMK if the pairwise ask is unclear though, that's the main thing.

gchanan commented:
for the generated file -- why do you need to be in the at::native namespace?  If everything is in an anonymous namespace anyway, it doesn't seem necessary.  If you are grabbing things from at:: or at::native::, you can just use "using" or write them out explicitly.

ezyang commented on aten/src/ATen/gen_backend_select_register.py:
```diff
@@ -0,0 +1,48 @@
+from code_template import CodeTemplate
```
It would be very nice if this file had a comment on top saying what it does

izdeby replied:
added a comment.

ezyang commented on aten/src/ATen/gen_backend_select_register.py:
```diff
@@ -0,0 +1,48 @@
+
+FUNCTION_DEFINITION = CodeTemplate("""\
+Tensor ${function_name}(${method_formals}) {
+  DispatchKey key = options.computeDispatchKey();
```
Temporary!! (no action needed)

ezyang commented on aten/src/ATen/gen_backend_select_register.py:
```diff
@@ -0,0 +1,48 @@
+                continue
+
+            name = option['name']
+            if 'overload_name' in option and option['overload_name'] != '':
```
nit: more compact way to test this is `option.get('overload_name')`

ezyang replied:
But I'm a little unsure why these need to be coded defensively as I was under the impression that `overload_name` and `dynamic_type` (below) were guaranteed to be set in the dictionaries.

izdeby replied:
its killed due to a new key 'backend_select' in the native_functions.yaml

gchanan replied:
I have the same question as @ezyang and I don't understand your response.

bhosmer replied:
...I'd still change this to `option.get('overload_name', '') != ''` if you have the chance

ezyang commented on c10/core/TensorOptions.h:
```diff
@@ -337,7 +337,7 @@ struct C10_API TensorOptions {
   // Resolves the tensor type set specified by the current construction axes.
   DispatchKeySet key_set() const noexcept {
-    return DispatchKeySet(computeDispatchKey());
+    return DispatchKeySet(computeDispatchKey()).add(DispatchKey::BackendSelect);
```
This can go in the global selector now that

izdeby replied:
what do you mean under global selector?

gchanan replied:
@ezyang: I think your comment got cut off.

My question would be: why do we add BackendSelect but not Variable?

ezyang replied:
```
  // Some keys are ALWAYS considered for inclusion by default, so they are
  // included in the set here.  (const appears to be sufficient for
  // always_included to get inlined, constexpr not necessary)
  const DispatchKeySet always_included{DispatchKey::VariableTensorId};
```

in `DispatchKeyExtractor.h`

This is why Variable doesn't need to be added here anymore, it's handled by `always_included`

gchanan replied:
right, so can we either unify those places or explain in a comment why they are different?

izdeby replied:
Added a comment.

ezyang reviewed (approved):
okey dokey

bhosmer reviewed (changes requested):
@gchanan had what I think is an important suggestion - I'm frontrunning it here :) basically, we should trigger this codegen by an explicit attribute in native_functions rather than by pattern matching. Inline comments to follow

bhosmer commented on aten/src/ATen/gen_backend_select_register.py:
```diff
@@ -0,0 +1,48 @@
+                name = "{0}_{1}".format(name, option['overload_name'])
+
+            if "arguments" in option and any("dynamic_type" in a and
+                                             a["dynamic_type"] == "TensorOptions" for a in option["arguments"]):
```
This kind of pattern matching logic - where we're empirically identifying a subset of ops by secondary characteristics, rather than via a) some explicit declaration or b) a higher-level concept or property of the system - is another thing I think we should try to dial out of the CG pipeline as we go, to make it more robust and discoverable. 

@gchanan 's suggestion here was to trigger the codegen by a `backend_select` attribute instead. I can't think of a downside to this - it makes the new capability an explicit futureproofed thing that can be documented, composed with other things, etc.

izdeby replied:
Added 'backend_select' key to native_function.yaml

izdeby committed:
Update on "Introduce BackendSelect dispatch key"

This stack is the first step toward an effort to fix, clean up and simplify code generation logic.  Please see the master [task](https://github.com/pytorch/pytorch/issues/30405) see related discussions and all the known issues.

The main focus of these changes is TensorOptions in code generation. 
Goals:
- Remove TensorOptions from generated code wherever it's possible. Leave it only in python/C++ API layers.
- Refactor TensorOptions logic to a single place.
- Log all discovered issues.

Non-goals:
- Fix Everything! 
- Remove all the hacks in code generation scripts.
- Clean up and defector all code generation scripts.

-----------
In this PR:
Introduce BackendSelect dispatch key with a high priority. All factory methods will be registering its special backend selector method with the ops schema so that the dispatcher would navigate there after Variable. Dispatch key id will be calculated and the correct kernel will be called. 

[Generated file](https://gist.github.com/izdeby/8056bf72b50b7c66075d37f334d84bf5)

-----------

Differential Revision: [D18912684](https://our.internmc.facebook.com/intern/diff/D18912684)

[ghstack-poisoned]

izdeby commented:
> What's here LGTM but I think I might not have been clear re colocating defs and registrations - I mean colocate them pairwise, so e.g. here's how `empty` and `empty_names` (to choose two random ops that are defined next to each other in the original, non-CG commit) would appear _[edit - added the per-op comment header to group things more clearly]_:
> 
> ```
> // in the middle of the list of op/options pairs:
> 
> // empty
> 
> Tensor empty(IntArrayRef size, const TensorOptions & options, c10::optional<MemoryFormat> memory_format) {
>   static auto op = c10::Dispatcher::singleton().findSchemaOrThrow("aten::empty", "memory_format");
>   DispatchKey key = options.computeDispatchKey();
>   return op.callUnboxedWithDispatchKey
>     <Tensor, IntArrayRef, const TensorOptions &, c10::optional<MemoryFormat>>
>     (key, size, options, memory_format);
> }
> 
> static auto empty_options = torch::RegisterOperators::options()
>     .schema("aten::empty.memory_format(int[] size, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None, MemoryFormat? memory_format=None) -> Tensor")
>     .impl_unboxedOnlyKernel<decltype(empty), &empty>(DispatchKey::BackendSelect)
>     .aliasAnalysis(AliasAnalysisKind::FROM_SCHEMA);
> 
> // empty_names
> 
> Tensor empty_names(IntArrayRef size, c10::optional<DimnameList> names, const TensorOptions & options, c10::optional<MemoryFormat> memory_format) {
>   return at::native::empty(size, names, options, memory_format);
> }
> 
> static auto empty_names_options = torch::RegisterOperators::options()
>     .schema("aten::empty.names(int[] size, *, Dimname[]? names, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None, MemoryFormat? memory_format=None) -> Tensor")
>     .impl_unboxedOnlyKernel<decltype(empty_names), &empty_names>(DispatchKey::BackendSelect)
>     .aliasAnalysis(c10::AliasAnalysisKind::FROM_SCHEMA);
> 
> // then a bunch more op/options pairs, then: 
> 
> static auto registry = torch::RegisterOperators()
>   ...
>   .op(empty_options)
>   .op(empty_names_options)
>   ...
> ```
> 
> ...there are other ways to group the pairs together, this was just a way that seemed readable to me. But the main thing is that all the nontrivial stuff related to op Foo appear in one place. (Even putting this example together drove the point home of how much friction it adds to have to go hunting around 😄)
> 
> Also FWIW, in this example I swapped the first two lines in the function def and added line breaks to the `callUnboxedWithDispatchKey` call - just some ideas for cheap ways to add readability.
> 
> LMK if the pairwise ask is unclear though, that's the main thing.

Sadly it can't be done as we don't have needed constructors for the torch::RegisterOperators::options(). as of today we are getting r-values and that wont work. We talked with @gchanan about this and it looks like it can't be done without substantial changes like introduction of a new constructor. we decided not to do that.

gchanan commented on c10/core/DispatchKeySet.h:
```diff
@@ -123,7 +123,7 @@ static inline DispatchKey legacyExtractDispatchKey(DispatchKeySet s) {
   // here.  At the moment, RequiresGrad (replacement for Variable)
   // is the most likely key that will need this treatment.
-  return s.highestPriorityTypeId();
+  return s.remove(DispatchKey::BackendSelect).highestPriorityTypeId();
```
perhaps you should update the comment to explain why BackendSelect is being removed.

izdeby replied:
Added a comment.

gchanan replied:
I know it's special -- but the question was why it's special in a different way than Variable (i.e. why BackendSelect and Variable aren't always inserted/removed in the same places).  Can you add a comment to that effect?

izdeby replied:
@ezyang, @gchanan does it make sense? Im having a hard time phrasing this.

gchanan commented on aten/src/ATen/templates/BackendSelectRegister.cpp:
```diff
@@ -0,0 +1,22 @@
+// This is used for factory methods dispatching. We register ops with a high priority dispatch key,
```
can you add `// ${generated_comment}` and make sure it gets filled in correctly -- look at other examples.

gchanan commented on aten/src/ATen/templates/BackendSelectRegister.cpp:
```diff
@@ -0,0 +1,22 @@
+// This is used for factory methods dispatching. We register ops with a high priority dispatch key,
+// which makes the dispatcher to dispatch here. We compute the dispatch key 'manually' and navigate to the
+// correct backend.
```
this description isn't clear at all.  There's no reason to be so general, you can give the specific examples.

For example:
"We register ops with a higher priority dispatch key (BackendSelect) than the usual backend-specific keys (e.g. CPUTensorId) which makes calls to the factory functions dispatch to here.  We then 'manually' compute a lower-priority to re-dispatch to (e.g. CPUTensorId) to get to the eventually correct backend."

gchanan commented on aten/src/ATen/native_parse.py:
```diff
@@ -413,6 +413,7 @@ def run(paths):
                 declaration['return'] = return_arguments if len(output_arguments) == 0 else output_arguments
                 declaration['variants'] = func.get('variants', ['function'])
                 declaration['requires_tensor'] = func.get('requires_tensor', False)
+                declaration['backend_select'] = func.get('backend_select', False)
```
I was actually making a different suggestion -- that BackendSelect is a dispatch key like any of the other ones, so it should be in the dispatch table, not it's own key.

gchanan commented on aten/src/ATen/templates/BackendSelectRegister.cpp:
```diff
@@ -0,0 +1,22 @@
+namespace at {
+namespace native {
+
+namespace {
```
did you respond to why this whole file can't be in an anonymous namespace?  (i.e. why do you need the at::native:: namespace).

gchanan replied:
I guess that's fine, but it's just namespace lookup that you can solve with a couple of "using" in the template.

izdeby committed:
Update on "[WIP: DO NOT REVIEW]Introduce BackendSelect dispatch key"

This stack is the first step toward an effort to fix, clean up and simplify code generation logic.  Please see the master [task](https://github.com/pytorch/pytorch/issues/30405) see related discussions and all the known issues.

The main focus of these changes is TensorOptions in code generation. 
Goals:
- Remove TensorOptions from generated code wherever it's possible. Leave it only in python/C++ API layers.
- Refactor TensorOptions logic to a single place.
- Log all discovered issues.

Non-goals:
- Fix Everything! 
- Remove all the hacks in code generation scripts.
- Clean up and defector all code generation scripts.

-----------
In this PR:
Introduce BackendSelect dispatch key with a high priority. All factory methods will be registering its special backend selector method with the ops schema so that the dispatcher would navigate there after Variable. Dispatch key id will be calculated and the correct kernel will be called. 

[Generated file](https://gist.github.com/izdeby/8056bf72b50b7c66075d37f334d84bf5)

-----------

Differential Revision: [D18912684](https://our.internmc.facebook.com/intern/diff/D18912684)

izdeby commented:
@gchanan, @ezyang, @bhosmer this should be good to go.
Things to note: 
1. to is being generated. Please, see the [generated file](https://gist.github.com/izdeby/8056bf72b50b7c66075d37f334d84bf5).
2. as discussed offline, we are not changing `native_functions.yaml` for now.

@bhosmer, some context for you.
1. We have discussed the idea of introducing the dispatch entry in `native_functions.yaml` for BeckendSelect but decided not to do that as its not a straightforward thing to do. As an example, take a look at 
`- func: randn(int[] size, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor`

it doesn't have any dispatch specified and if we will add one for the BeckendSelect, the current logic requires us to specify all other cases like for CPU/CUDA.

It wasn't obvious what is the right thing to do here so we postponed it for now.

2. Regarding reordering the definitions and registrations in generated files - please, see the comment above but long story short, it cant be easily done. we would need to introduce additional registration mechanisms and that's a very different change.

bhosmer commented on aten/src/ATen/gen_backend_select_register.py:
```diff
@@ -0,0 +1,56 @@
+            if 'overload_name' in option and option['overload_name'] != '':
+                name = "{0}_{1}".format(name, option['overload_name'])
+
+            if any("dynamic_type" in a and a["dynamic_type"] == "TensorOptions" for a in option["arguments"]):
```
Thanks for the update on why you guys decided to stick with the pattern matching trigger for now, makes sense to keep the scope of the PR manageable. 

My only suggestion would be to clarify what's going on by factoring the triggering condition out into its own function, i.e. something like (note the ... in the comment, where I don't know the exact reasoning but it'd be good to have it here):
```
def needs_backend_select(decl):
    # currently, we register an op under the BackendSelect dispatch key
    # if a TensorOptions argument has been gathered from its declared args
    # (excluding some ops for ... reasons). 

    if '_like' in decl['name'] or 'new_' in decl['name']:
        return false
    return any(a.get('dynamic_type') == 'TensorOptions' for a in decl('arguments'))
```

izdeby replied:
done!

bhosmer reviewed (approved):
Hey @izdeby, thanks for the context! Makes sense to me - accepting to unblock, but see the factoring suggestion inline - I definitely think that would be worth doing if it's not too inconvenient.

izdeby committed:
Update on "Introduce BackendSelect dispatch key"

This stack is the first step toward an effort to fix, clean up and simplify code generation logic.  Please see the master [task](https://github.com/pytorch/pytorch/issues/30405) see related discussions and all the known issues.

The main focus of these changes is TensorOptions in code generation. 
Goals:
- Remove TensorOptions from generated code wherever it's possible. Leave it only in python/C++ API layers.
- Refactor TensorOptions logic to a single place.
- Log all discovered issues.

Non-goals:
- Fix Everything! 
- Remove all the hacks in code generation scripts.
- Clean up and defector all code generation scripts.

-----------
In this PR:
Introduce BackendSelect dispatch key with a high priority. All factory methods will be registering its special backend selector method with the ops schema so that the dispatcher would navigate there after Variable. Dispatch key id will be calculated and the correct kernel will be called. 

[Generated file](https://gist.github.com/izdeby/8056bf72b50b7c66075d37f334d84bf5)

-----------

Differential Revision: [D18912684](https://our.internmc.facebook.com/intern/diff/D18912684)

[ghstack-poisoned]

gchanan commented on aten/src/ATen/gen_backend_select_register.py:
```diff
@@ -0,0 +1,64 @@
+from code_template import CodeTemplate
+
+GENERATED_COMMENT = CodeTemplate(
+    "@" + "generated from ${filename}")
```
the other examples of this do this via an entry in the template, right?

izdeby replied:
Yeah, its being [set once](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/gen.py#L118) and used everywhere.

gchanan replied:
I meant in the template file, not in the codegen.  I don't see any reason for this to be in the codegen.  See e.g: https://github.com/pytorch/pytorch/blob/8195961f20caa18863698d85410d0818bfd67463/aten/src/ATen/templates/Functions.h#L3

gchanan replied:
ah, I guess you aren't passing around the env (from the top-layer) so you can't use it.

gchanan commented on aten/src/ATen/gen_backend_select_register.py:
```diff
@@ -0,0 +1,64 @@
+    # if a TensorOptions argument has been gathered from its declared args
+    # We skip all the 'new_*' and '*_like' ops as they are special cased and avoid dispatching.
+    # See TypeDefault.cpp
+    if '_like' in declaration_option['name'] or 'new_' in declaration_option['name']:
```
nit: can we use startswith / endswith?

izdeby replied:
Done

gchanan commented on aten/src/ATen/gen_backend_select_register.py:
```diff
@@ -0,0 +1,64 @@
+    if '_like' in declaration_option['name'] or 'new_' in declaration_option['name']:
+        return False
+
+    return any("dynamic_type" in a and a["dynamic_type"] == "TensorOptions" for a in declaration_option["arguments"])
```
it's more idiomatic to do something like:
`return any(a.get('dynamic_type') == 'TensorOptions' for a in declaration_option['arguments'])`

izdeby replied:
Updated

izdeby committed:
Update on "Introduce BackendSelect dispatch key"

This stack is the first step toward an effort to fix, clean up and simplify code generation logic.  Please see the master [task](https://github.com/pytorch/pytorch/issues/30405) see related discussions and all the known issues.

The main focus of these changes is TensorOptions in code generation. 
Goals:
- Remove TensorOptions from generated code wherever it's possible. Leave it only in python/C++ API layers.
- Refactor TensorOptions logic to a single place.
- Log all discovered issues.

Non-goals:
- Fix Everything! 
- Remove all the hacks in code generation scripts.
- Clean up and defector all code generation scripts.

-----------
In this PR:
Introduce BackendSelect dispatch key with a high priority. All factory methods will be registering its special backend selector method with the ops schema so that the dispatcher would navigate there after Variable. Dispatch key id will be calculated and the correct kernel will be called. 

[Generated file](https://gist.github.com/izdeby/8056bf72b50b7c66075d37f334d84bf5)

-----------

Differential Revision: [D18912684](https://our.internmc.facebook.com/intern/diff/D18912684)

[ghstack-poisoned]

gchanan commented on aten/src/ATen/gen_backend_select_register.py:
```diff
@@ -0,0 +1,64 @@
+                if option.get('overload_name', '') != '':
+                    name = "{0}_{1}".format(name, option['overload_name'])
+
+                func_reg = FUNCTION_REGISTRATION.substitute(schema_string=option['schema_string'],
```
this code assumes that we are registering an unboxed kernel, but it's not checked.  Can we add an `assert option['use_c10_dispatcher'] == 'unboxed_only'`

izdeby replied:
added

gchanan commented on c10/core/DispatchKeySet.h:
```diff
@@ -123,7 +123,14 @@ static inline DispatchKey legacyExtractDispatchKey(DispatchKeySet s) {
+  // It was introduced only for the factory functions with TensorOptions. The logic that
+  // uses legacyExtractDispatchKey doesnt expect BackendSelect and VariableTensorId keys.
+  // VariableTensorId is being excluded from a DispatchKeySet right after dispatching
+  // (See variable_excluded_from_dispatch in TensorBody.h)
```
that is just an example where we are checking if something was historically a variable -- this is not the place where variable is being excluded.

izdeby committed:
Update on "Introduce BackendSelect dispatch key"

This stack is the first step toward an effort to fix, clean up and simplify code generation logic.  Please see the master [task](https://github.com/pytorch/pytorch/issues/30405) see related discussions and all the known issues.

The main focus of these changes is TensorOptions in code generation. 
Goals:
- Remove TensorOptions from generated code wherever it's possible. Leave it only in python/C++ API layers.
- Refactor TensorOptions logic to a single place.
- Log all discovered issues.

Non-goals:
- Fix Everything! 
- Remove all the hacks in code generation scripts.
- Clean up and defector all code generation scripts.

-----------
In this PR:
Introduce BackendSelect dispatch key with a high priority. All factory methods will be registering its special backend selector method with the ops schema so that the dispatcher would navigate there after Variable. Dispatch key id will be calculated and the correct kernel will be called. 

[Generated file](https://gist.github.com/izdeby/8056bf72b50b7c66075d37f334d84bf5)

-----------

Differential Revision: [D18912684](https://our.internmc.facebook.com/intern/diff/D18912684)

[ghstack-poisoned]

izdeby committed:
Update on "Introduce BackendSelect dispatch key"

This stack is the first step toward an effort to fix, clean up and simplify code generation logic.  Please see the master [task](https://github.com/pytorch/pytorch/issues/30405) see related discussions and all the known issues.

The main focus of these changes is TensorOptions in code generation. 
Goals:
- Remove TensorOptions from generated code wherever it's possible. Leave it only in python/C++ API layers.
- Refactor TensorOptions logic to a single place.
- Log all discovered issues.

Non-goals:
- Fix Everything! 
- Remove all the hacks in code generation scripts.
- Clean up and defector all code generation scripts.

-----------
In this PR:
Introduce BackendSelect dispatch key with a high priority. All factory methods will be registering its special backend selector method with the ops schema so that the dispatcher would navigate there after Variable. Dispatch key id will be calculated and the correct kernel will be called. 

[Generated file](https://gist.github.com/izdeby/8056bf72b50b7c66075d37f334d84bf5)

-----------

Differential Revision: [D18912684](https://our.internmc.facebook.com/intern/diff/D18912684)

[ghstack-poisoned]

izdeby committed:
Update on "Introduce BackendSelect dispatch key"

This stack is the first step toward an effort to fix, clean up and simplify code generation logic.  Please see the master [task](https://github.com/pytorch/pytorch/issues/30405) see related discussions and all the known issues.

The main focus of these changes is TensorOptions in code generation. 
Goals:
- Remove TensorOptions from generated code wherever it's possible. Leave it only in python/C++ API layers.
- Refactor TensorOptions logic to a single place.
- Log all discovered issues.

Non-goals:
- Fix Everything! 
- Remove all the hacks in code generation scripts.
- Clean up and defector all code generation scripts.

-----------
In this PR:
Introduce BackendSelect dispatch key with a high priority. All factory methods will be registering its special backend selector method with the ops schema so that the dispatcher would navigate there after Variable. Dispatch key id will be calculated and the correct kernel will be called. 

[Generated file](https://gist.github.com/izdeby/8056bf72b50b7c66075d37f334d84bf5)

-----------

Differential Revision: [D18912684](https://our.internmc.facebook.com/intern/diff/D18912684)

[ghstack-poisoned]

izdeby committed:
Update on "Introduce BackendSelect dispatch key"

This stack is the first step toward an effort to fix, clean up and simplify code generation logic.  Please see the master [task](https://github.com/pytorch/pytorch/issues/30405) see related discussions and all the known issues.

The main focus of these changes is TensorOptions in code generation. 
Goals:
- Remove TensorOptions from generated code wherever it's possible. Leave it only in python/C++ API layers.
- Refactor TensorOptions logic to a single place.
- Log all discovered issues.

Non-goals:
- Fix Everything! 
- Remove all the hacks in code generation scripts.
- Clean up and defector all code generation scripts.

-----------
In this PR:
Introduce BackendSelect dispatch key with a high priority. All factory methods will be registering its special backend selector method with the ops schema so that the dispatcher would navigate there after Variable. Dispatch key id will be calculated and the correct kernel will be called. 

[Generated file](https://gist.github.com/izdeby/8056bf72b50b7c66075d37f334d84bf5)

-----------

Differential Revision: [D18912684](https://our.internmc.facebook.com/intern/diff/D18912684)

[ghstack-poisoned]

izdeby committed:
Update on "Introduce BackendSelect dispatch key"

This stack is the first step toward an effort to fix, clean up and simplify code generation logic.  Please see the master [task](https://github.com/pytorch/pytorch/issues/30405) see related discussions and all the known issues.

The main focus of these changes is TensorOptions in code generation. 
Goals:
- Remove TensorOptions from generated code wherever it's possible. Leave it only in python/C++ API layers.
- Refactor TensorOptions logic to a single place.
- Log all discovered issues.

Non-goals:
- Fix Everything! 
- Remove all the hacks in code generation scripts.
- Clean up and defector all code generation scripts.

-----------
In this PR:
Introduce BackendSelect dispatch key with a high priority. All factory methods will be registering its special backend selector method with the ops schema so that the dispatcher would navigate there after Variable. Dispatch key id will be calculated and the correct kernel will be called. 

[Generated file](https://gist.github.com/izdeby/8056bf72b50b7c66075d37f334d84bf5)

-----------

Differential Revision: [D18912684](https://our.internmc.facebook.com/intern/diff/D18912684)

[ghstack-poisoned]
